torch/_numpy/_util.py

```diff
@@ -220,7 +228,7 @@ def _coerce_to_tensor(obj, dtype=None, copy=False, ndmin=0):
         # Therefore, we treat `tensor.dtype` as a hint, and convert the
         # original object *again*, this time with an explicit dtype.
         torch_dtype = _dtypes_impl.get_default_dtype_for(tensor.dtype)
-        tensor = torch.as_tensor(obj, dtype=torch_dtype)
+        tensor = _try_convert_to_tensor(obj, dtype=torch_dtype)
```

Comment:
Not for this PR:
I really dislike this pattern, as you know. We should have a happy  path that's hit pretty much 100% of the time, and have this as a last resort for the very weird edge cases.

Reply:
There are two patterns we both dislike: the need to call as_tensor twice because dtypes and an EAFP. If the comment is about the former, I don't see a way out; if it's about the latter, https://github.com/pytorch/pytorch/pull/112521#discussion_r1378065685 ¯\_(ツ)_/¯

Reply:
If there's no scalar, we can just call it once. That pretty much covers 99% of the uses of this function, I reckon.

Reply:
There's no scalar recursively as in, everythin is a tensor, or an ndarray or, recursively, a list of these.

Reply:
Calling as_tensor twice is about dtypes, not shapes.

Reply:
Actually, nevermind, we can implement this with one `as_tensor` using https://pytorch.org/docs/stable/generated/torch.set_default_dtype.html and then rolling back the previous default.